adam/learner/affordances.py

```diff
@@ -0,0 +1,231 @@
+                            surface_template=SurfaceTemplate(
+                                elements=(slot.name, action_node.concept.debug_string),
+                                language_mode=self._language_mode,
+                            ),
+                            slot_to_semantic_node={SLOT1: object_node},
+                        )
+                    )
+
+        return immutableset(candidates)
+
+    def _enrich_post_process(
+        self,
+        perception_graph_after_matching: PerceptionGraph,
+        immutable_new_nodes: AbstractSet[SemanticNode],
+    ) -> Tuple[PerceptionGraph, AbstractSet[SemanticNode]]:
+
+        digraph = perception_graph_after_matching.copy_as_digraph()
+        for node in immutable_new_nodes:
+            if (
+                isinstance(node, AffordanceSemanticNode)
+                and len(node.slot_fillings.values()) == 1
+            ):
+                # Ideally I'd like this weight to be the graph match percentage, but we don't have that information
+                # Available to us here and refactoring to have that information is out of scope
+                affordance_property = CategoricalNode(
+                    label="affordance", value=node.concept.debug_string, weight=1.0
+                )
+                perception_graph_after_matching = add_node_connected_to_perception_graph(
+                    perception_graph_after_matching,
+                    list(node.slot_fillings.values())[0],
+                    affordance_property,
+                    temporal_scope=TemporalScope.BEFORE
+                    if perception_graph_after_matching.dynamic
+                    else None,
+                )
+
+        return (
+            PerceptionGraph(
+                graph=digraph, dynamic=perception_graph_after_matching.dynamic
+            ),
+            immutable_new_nodes,
+        )
+
+
+@attrs(slots=True)
+class SubsetAffordanceLearner(
+    AbstractAffordanceTemplateLearner, AbstractTemplateSubsetLearner
+):
+    """
+    An implementation of `TopLevelLanguageLearner` for subset learning based approach for affordance learning.
+    """
+
+    def _update_hypothesis(
+        self,
+        previous_pattern_hypothesis: PerceptionGraphTemplate,
+        current_pattern_hypothesis: PerceptionGraphTemplate,
+    ) -> Optional[PerceptionGraphTemplate]:
+        return previous_pattern_hypothesis.intersection(
+            current_pattern_hypothesis,
+            ontology=self._ontology,
+            match_mode=MatchMode.NON_OBJECT,
+            allowed_matches=immutablesetmultidict(
+                [
+                    (node2, node1)
+                    for previous_slot, node1 in previous_pattern_hypothesis.template_variable_to_pattern_node.items()
+                    for new_slot, node2 in current_pattern_hypothesis.template_variable_to_pattern_node.items()
+                    if previous_slot == new_slot
+                ]
+            ),
+        )
```

Comment:
This need to be updated similarly to how the other subset object learners were updated. See 2ad58ac5ff9e9e9bbd86aeeb7f36cb158e82ca02.